polkadot/node/core/consensus-statistics-collector/src/lib.rs

```diff
@@ -0,0 +1,393 @@
+                    let relay_hash = activated.hash;
+
+                    let (tx, rx) = oneshot::channel();
+
+                    ctx.send_message(ChainApiMessage::BlockHeader(relay_hash, tx)).await;
+                    let header = rx
+                        .map_err(JfyiError::OverseerCommunication)
+                        .await?
+                        .map_err(JfyiError::ChainApiCallError)?;
+
+                    let session_idx = request_session_index_for_child(relay_hash, ctx.sender())
+                        .await
+                        .await
+                        .map_err(JfyiError::OverseerCommunication)?
+                        .map_err(JfyiError::RuntimeApiCallError)?;
+
+                    if let Some(ref h) = header {
+                        let parent_hash = h.clone().parent_hash;
+                        let parent_hash = if let Some(parent) = view.per_relay.get_mut(&parent_hash) {
+                            parent.link_child(relay_hash.clone());
+                            Some(parent_hash)
+                        } else {
+                            view.roots.insert(relay_hash.clone());
+                            None
+                        };
+
+                        view.per_relay.insert(relay_hash, PerRelayView::new(parent_hash, session_idx));
+                    } else {
+                        view.roots.insert(relay_hash.clone());
+                        view.per_relay.insert(relay_hash, PerRelayView::new(None, session_idx));
+                    }
+
+                    if !view.per_session.contains_key(&session_idx) {
+                        let session_info = request_session_info(relay_hash, session_idx, ctx.sender())
+                            .await
+                            .await
+                            .map_err(JfyiError::OverseerCommunication)?
+                            .map_err(JfyiError::RuntimeApiCallError)?;
+
+                        if let Some(session_info) = session_info {
+                            let mut authority_lookup = HashMap::new();
+                            for (i, ad) in session_info.discovery_keys.iter().cloned().enumerate() {
+                                authority_lookup.insert(ad, ValidatorIndex(i as _));
+                            }
+
+                            view.per_session.insert(session_idx, PerSessionView::new(authority_lookup));
+                        }
+                    }
+                }
+            },
+            FromOrchestra::Signal(OverseerSignal::BlockFinalized(fin_block_hash, _)) => {
+                // check if a session was finalized
+                let session_idx = request_session_index_for_child(fin_block_hash, ctx.sender())
+                    .await
+                    .await
+                    .map_err(JfyiError::OverseerCommunication)?
+                    .map_err(JfyiError::RuntimeApiCallError)?;
+
+                let should_prune = match view.current_session {
```

Comment:
no need to keep unfinalized block for so long around you can simply, prune them if they aren't on the finalized chain.